kolibri/core/assets/src/exams/utils.js

```diff
@@ -84,106 +88,110 @@ function annotateQuestionsWithItem(questions) {
-    contentNodes.forEach(node => {
-      questionIds[node.id] = node.assessmentmetadata
-        ? node.assessmentmetadata.assessment_item_ids
-        : [];
+export async function convertExamQuestionSources(exam) {
+  if (exam.data_model_version === 0) {
+    const ids = uniq(exam.question_sources.map(item => item.exercise_id));
+    const exercises = await ContentNodeResource.fetchCollection({
+      getParams: {
+        ids,
+        no_available_filtering: true,
+      },
     });
-    return annotateQuestionsWithItem(
-      convertExamQuestionSourcesV0V2(exam.question_sources, exam.seed, questionIds)
+    const questionIds = exercises.reduce((nodeIds, node) => {
+      nodeIds[node.id] = node.assessmentmetadata ? node.assessmentmetadata.assessment_item_ids : [];
+      return nodeIds;
+    }, []);
+    exam.question_sources = convertExamQuestionSourcesV0V2(
+      exam.question_sources,
+      exam.seed,
+      questionIds
     );
+    // v1 -> v2 only updates the `counter_in_exercise` field if it's in camelCase
+    // so we can set the data_model_version to 2 here to skip that code
+    exam.data_model_version = 2;
   }
-  if (data_model_version === 1) {
-    return annotateQuestionsWithItem(convertExamQuestionSourcesV1V2(exam.question_sources));
+
+  if (exam.data_model_version === 1) {
+    exam.question_sources = convertExamQuestionSourcesV1V2(exam.question_sources);
+    exam.data_model_version = 2;
   }
 
-  // For backwards compatibility. If you are using V3, use the convertExamQuestionSourcesToV3 func
-  if (data_model_version === 3) {
-    return revertV3toV2(exam.question_sources);
+  if (exam.data_model_version === 2) {
+    exam.question_sources = convertExamQuestionSourcesV2toV3(exam);
+    exam.data_model_version = 3;
   }
 
-  return annotateQuestionsWithItem(exam.question_sources);
+  // Now we know we have the latest V3 structure
+  exam.question_sources = exam.question_sources.map(section => {
+    section.questions = annotateQuestionsWithItem(section.questions);
+    return section;
+  });
+
+  return exam;
 }
 
-export function fetchNodeDataAndConvertExam(exam) {
-  const { data_model_version } = exam;
-  if (data_model_version >= 3) {
-    /* For backwards compatibility, we need to convert V3 to V2 */
-    exam.question_sources = revertV3toV2(exam.question_sources);
-    return Promise.resolve(exam);
-  }
-  if (data_model_version == 2) {
-    exam.question_sources = annotateQuestionsWithItem(exam.question_sources);
-    return Promise.resolve(exam);
-  }
-  return ContentNodeResource.fetchCollection({
-    getParams: {
-      ids: uniq(exam.question_sources.map(item => item.exercise_id)),
-      no_available_filtering: true,
-    },
-  }).then(contentNodes => {
-    return {
-      ...exam,
-      question_sources: convertExamQuestionSources(exam, { contentNodes }),
-    };
+/**
+ * @returns {Promise} - resolves to an object with the exam and the exercises
+ */
+export async function fetchExamWithContent(exam) {
```

Comment:
This function name is mildly misleading, as it suggests this function fetches the exam, whereas it actually requires the exam to already have been fetched and passed in as the only argument.